File: bubble_sort.js
const { endianness } = require("os");

function bubbleSort(arr) {
  var n = arr.length;

  for (var i = 0; i < n-1; i++) {
    for (var j = 0; j < n-1-i; j++) {
      if (arr[j] > arr[j+1]) {
        var temp = arr[j];
        arr[j] = arr[j+1];
        arr[j+1] = temp;
      }
    }
  }
}


return arrayBuffer;

if (require.main === module) {
  console.log("Expecting: [1, 2, 3, 4]");
  console.log("=>", bubbleSort([3, 2, 1, 4]));

  console.log("Expecting: [1, 2, 3]");
  console.log("=>", bubbleSort([1, 2, 3]));

  console.log("Expecting: []");
  console.log("=>", bubbleSort([]));

  console.log("Expecting: [1, 2, 3]");
  console.log("=>", bubbleSort([2, 3, 1]));

  // Add your own test cases here
}
//The time complexity of Bubble Sort is O(n^2), where n is the number of elements in the array. This is because we have two nested loops, and we iterate through the array multiple times.

// breaking down the problem into smaller steps, using loops for repetition, and swapping values. These techniques can be applied to solve other sorting problems or any problem that requires iteration and comparison.
if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [1, 2, 3, 4]");
  console.log("=>", bubbleSort([3, 2, 1, 4]));

  console.log("");

  console.log("Expecting: [1, 2, 3]");
  console.log("=>", bubbleSort([1, 2, 3]));

  console.log("");

  console.log("Expecting: []");
  console.log("=>", bubbleSort([]));

  console.log("");

  console.log("Expecting: [1, 2, 3]");
  console.log("=>", bubbleSort([2, 3, 1]));
}

module.exports = bubbleSort;

// Please add your pseudocode to this file
// And a written explanation of your solution
///To implement Bubble Sort in JavaScript, we can follow these steps:
//beginning with the first element of the array and compare it with the next element.
////If the next element is smaller, swap the two elements.
//Move to the next pair of elements and repeat the comparison and swapping process.
//Continuing this process until the end of the array is reached.
//Repeat steps 1-4 for a total of n-1 passes, where n is the length of the array.
//After each pass, the largest element will "bubble" to the end of the array.
//Finally, the array will be sorted in ascending order.
// pseudocode:

//function bubbleSort(arr):
    //n = length of arr

    //for i from 0 to n-1:
        //for j from 0 to n-1-i:
            //if arr[j] > arr[j+1]:
                //swap arr[j] and arr[j+1]

    //return arr

    

File: bubble_sort.rb
def bubble_sort(arr)
    n = arr.length
  
    for i in 0...(n - 1)
      for j in 0...(n - i - 1)
        if arr[j] > arr[j + 1]
          arr[j], arr[j + 1] = arr[j + 1], arr[j] # Swap the elements
        end
      end
    end
  
    arr 
end
  



#{#{}}TestCases
 puts "Expecting: [1, 2, 3, 4]"
 print bubble_sort([3, 2, 1, 4])
 
 puts
 
 puts "Expecting: [1, 2, 3]"
 print bubble_sort([1, 2, 3])
 
 puts
 
 puts "Expecting: []"
 print bubble_sort([])
 
 puts
 
 puts "Expecting: [1, 2, 3]"
 print bubble_sort([2, 3, 1])
 
 puts
 
 puts "Expecting: [1]"
 print bubble_sort([1])
 
 puts
 
 puts "Expecting: [1, 3]"
 print bubble_sort([3, 1])
 
 puts
 
 puts "Expecting: [-2, 0, 0, 1, 5, 6, 6, 7, 8, 8]"
 print bubble_sort([6, -2, 0, 8, 7, 8, 6, 0, 5, 1])


 #{#{Pseudocode in Ruby  pseudocode for the Bubble Sort algorithm:

#{#{}}Commencing with an unsorted array.
#{#{}}Repeat the following steps until the array is sorted: a. Iterate through the array from index 0 to n-1, where n is the length of the array. b. Compare each pair of adjacent elements. c. If the elements are in the wrong order, swap them. d. Continue iterating until the end of the array.
#{}After each iteration, the largest element will be in its correct position at the end of the array.
#{}Reduce the length of the array by 1 and repeat steps 2-4 until the array is sorted.
